File: tests/parity/run_parity.ts
/**
 * Cross-SDK billing parity runner — JavaScript/TypeScript side.
 *
 * Loads billing_scenarios.json and exercises each scenario against the JS
 * MemoryStore + CreditManager. Results are compared field-by-field against the
 * expected values in the JSON. Any field-level divergence is a test failure.
 *
 * Run via: cd javascript && npx vitest run ../tests/parity/run_parity.ts
 */
import { describe, it, expect } from "vitest";
import { readFileSync } from "fs";
import { resolve, dirname } from "path";
import { fileURLToPath } from "url";
import { CreditManager } from "../../javascript/src/manager.js";
import { MemoryStore } from "../../javascript/src/stores/memory-store.js";
import { evaluateExpression } from "../../javascript/src/expr.js";
import Decimal from "decimal.js";
import {
  InsufficientCreditsError,
  CapReachedError,
  LeaseExpiredError,
  LeaseNotFoundError,
} from "../../javascript/src/errors.js";
import type {
  PricingConfigData,
} from "../../javascript/src/types.js";

// ---------------------------------------------------------------------------
// Load scenario fixture
// ---------------------------------------------------------------------------

const __filename = fileURLToPath(import.meta.url);
const __dir = dirname(__filename);
const SCENARIOS_FILE = resolve(__dir, "billing_scenarios.json");
const _data = JSON.parse(readFileSync(SCENARIOS_FILE, "utf-8"));
const PRICING: PricingConfigData = _data.pricing_config;
const SCENARIOS: Array<Record<string, unknown>> = _data.scenarios;

const D = (v: string | number) => new Decimal(v);

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function q(v: Decimal): string {
  return v.toDecimalPlaces(4, Decimal.ROUND_HALF_UP).toFixed(4);
}

function errorCode(err: unknown): string {
  if (err instanceof Error) {
    // Check if the message embeds a known code.
    const knownCodes = [
      "spend_cap_exceeded",
      "cap_reached",
      "insufficient_team_balance",
      "lease_expired",
      "lease_not_found",
      "invalid_amount",
    ];
    for (const code of knownCodes) {
      if (err.message.includes(code)) return code;
    }
    if (err instanceof InsufficientCreditsError) return "insufficient_credits";
    if (err instanceof CapReachedError) return "cap_reached";
    if (err instanceof LeaseExpiredError) return "lease_expired";
    if (err instanceof LeaseNotFoundError) return "lease_not_found";
    return err.constructor.name;
  }
  return String(err);
}

function assertFields(
  scenarioName: string,
  expected: Record<string, unknown>,
  got: Record<string, unknown>,
): void {
  const failures: string[] = [];
  for (const [key, expVal] of Object.entries(expected)) {
    const gotVal = got[key];
    // Numeric string comparison: normalise both sides to 4 d.p.
    if (typeof expVal === "string" && expVal.includes(".")) {
      try {
        const expD = D(expVal);
        const gotD = gotVal != null ? D(String(gotVal)) : null;
        if (gotD === null || !expD.eq(gotD)) {
          failures.push(`  ${key}: expected '${expVal}', got '${String(gotVal)}'`);
        }
        continue;
      } catch {
        // fall through to strict equality
      }
    }
    if (expVal !== gotVal) {
      failures.push(`  ${key}: expected ${JSON.stringify(expVal)}, got ${JSON.stringify(gotVal)}`);
    }
  }
  if (failures.length > 0) {
    throw new Error(
      `Scenario '${scenarioName}' failed:\n${failures.join("\n")}\n  result: ${JSON.stringify(got)}`,
    );
  }
}

// ---------------------------------------------------------------------------
// Setup helpers
// ---------------------------------------------------------------------------

async function runSetup(
  store: MemoryStore,
  steps: Array<Record<string, unknown>>,
): Promise<void> {
  for (const step of steps) {
    const op = step["op"] as string;
    if (op === "add_credits") {
      const amt = D(step["amount"] as string);
      // Skip zero-amount adds (just ensure the user has a balance entry of 0).
      if (amt.isZero()) {
        // Force a balance entry to exist so the user is "known" but has 0 balance.
        // addCredits would reject 0; for zero-balance overdraft tests we just skip.
        return;
      }
      await store.addCredits(step["user_id"] as string, amt);
    } else if (op === "set_user_plan") {
      await store.setUserPlan(step["user_id"] as string, step["plan_key"] as string);
    } else if (op === "create_team") {
      // MemoryStore generates a UUID; patch internal maps to stable fixture id.
      // We use `as any` because `teams` / `teamMembers` are private.
      const result = await store.createTeam(
        step["team_id"] as string,
        D(step["initial_balance"] as string),
      );
      const actualId = result.teamId;
      const fixtureId = step["team_id"] as string;
      if (actualId !== fixtureId) {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const s = store as any;
        const teamRec = s.teams.get(actualId)!;
        s.teams.delete(actualId);
        teamRec.id = fixtureId;
        s.teams.set(fixtureId, teamRec);
        const memberMap = s.teamMembers.get(actualId)!;
        s.teamMembers.delete(actualId);
        s.teamMembers.set(fixtureId, memberMap);
      }
    } else if (op === "add_team_member") {
      const cap = step["spend_cap"] ? D(step["spend_cap"] as string) : undefined;
      await store.addTeamMember(
        step["team_id"] as string,
        step["user_id"] as string,
        "member",
        cap,
      );
    } else {
      throw new Error(`Unknown setup op: ${op}`);
    }
  }
}

// ---------------------------------------------------------------------------
// Action runner
// ---------------------------------------------------------------------------

async function runAction(
  scenario: Record<string, unknown>,
  pricing: PricingConfigData,
): Promise<Record<string, unknown>> {
  const action = scenario["action"] as Record<string, unknown>;
  const op = action["op"] as string;

  // Mirror Python runner: create manager with billing mode/floor from action.
  const actionBillingMode = (action["billing_mode"] as string | undefined) ?? "strict";
  const actionOverdraftFloorRaw = action["overdraft_floor"] as string | number | undefined;
  // Pass overdraftFloor as a number to avoid Decimal instanceof cross-module issues.
  const actionOverdraftFloorNum = actionOverdraftFloorRaw != null
    ? parseFloat(String(actionOverdraftFloorRaw))
    : undefined;
  const policy = actionBillingMode === "overdraft" ? "overdraft" : "strict_prepaid";

  const store = new MemoryStore();
  const manager = new CreditManager(store, undefined, undefined, {
    policy,
    ...(actionOverdraftFloorNum !== undefined ? { overdraftFloor: actionOverdraftFloorNum } : {}),
  });
  await manager.publishPricingFromDict(pricing);

  await runSetup(store, (scenario["setup"] as Array<Record<string, unknown>>) ?? []);

  if (op === "deduct" || op === "deduct_idempotent") {
    const metrics = {
      model: (action["model"] as string | undefined) ?? "_default",
      inputTokens: (action["input_tokens"] as number | undefined) ?? 0,
      outputTokens: (action["output_tokens"] as number | undefined) ?? 0,
    };
    try {
      if (op === "deduct_idempotent") {
        const ikey = action["idempotency_key"] as string;
        const r1 = await manager.deduct(action["user_id"] as string, metrics, ikey);
        const r2 = await manager.deduct(action["user_id"] as string, metrics, ikey);
        return {
          balance_after: q(r1.balanceAfter),
          idempotent_on_replay: r2.idempotent,
          error: null,
        };
      } else {
        const r = await manager.deduct(action["user_id"] as string, metrics);
        return {
          balance_after: q(r.balanceAfter),
          allowance_consumed: r.allowanceConsumed.toFixed(),
          plan_covered: r.planCovered ?? null,
          error: r.error ?? null,
        };
      }
    } catch (e) {
      if (process.env["PARITY_DEBUG"]) {
        console.error(`[${op}] error:`, e);
        if (e instanceof Error) console.error("stack:", e.stack?.split("\n").slice(0,5).join("\n"));
      }
      return { error: errorCode(e) };
    }
  }

  if (op === "get_user_plan") {
    const r = await store.getUserPlan(action["user_id"] as string);
    return {
      plan_id: r.planId,
      billing_mode: r.defaultBillingMode ?? "strict",
    };
  }

  if (op === "insufficient_credits_strict") {
    // handled via deduct path — but this scenario uses "deduct" op
    return {};
  }

  if (op === "lease_reserve_settle") {
    const userId = action["user_id"] as string;
    // Use numbers to avoid cross-module Decimal instanceof issues with isAmount().
    const reserveAmt = parseFloat(action["reserve_amount"] as string);
    const settleAmtDecimal = D(action["settle_amount"] as string);
    const minBalanceDecimal = action["min_balance"]
      ? D(action["min_balance"] as string)
      : undefined;
    try {
      if (process.env["PARITY_DEBUG"]) {
        const bal = await store.getBalance(userId);
        console.log(`[lease] userId=${userId} balance=${bal.balance} reserveAmt=${reserveAmt} policy=${policy} floor=${actionOverdraftFloorNum}`);
      }
      const lease = await manager.reserve(userId, reserveAmt);
      const r = await store.settleLease(userId, lease.leaseId, settleAmtDecimal, {
        minBalance: minBalanceDecimal ?? new Decimal(0),
      });
      return {
        balance_after: q(r.balanceAfter),
        error: r.error ?? null,
      };
    } catch (e) {
      if (process.env["PARITY_DEBUG"]) console.error(`[lease_reserve_settle] error:`, e);
      return { error: errorCode(e) };
    }
  }

  if (op === "lease_reserve_release") {
    const userId = action["user_id"] as string;
    const reserveAmt = parseFloat(action["reserve_amount"] as string);
    try {
      const lease = await manager.reserve(userId, reserveAmt);
      await manager.release(userId, lease.leaseId);
      const avail = await manager.getAvailable(userId);
      return {
        available_after_release: q(avail.available),
        error: null,
      };
    } catch (e) {
      return { error: errorCode(e) };
    }
  }

  if (op === "team_deduct") {
    const metrics = {
      model: (action["model"] as string | undefined) ?? "_default",
      inputTokens: (action["input_tokens"] as number | undefined) ?? 0,
      outputTokens: (action["output_tokens"] as number | undefined) ?? 0,
    };
    try {
      const r = await manager.deductTeam(
        action["team_id"] as string,
        action["user_id"] as string,
        metrics,
      );
      return {
        team_balance_after: q(r.teamBalanceAfter),
        error: r.error ?? null,
      };
    } catch (e) {
      return { error: errorCode(e) };
    }
  }

  if (op === "team_deduct_twice") {
    const metrics1 = {
      model: "_default",
      inputTokens: (action["input_tokens_1"] as number | undefined) ?? 0,
      outputTokens: (action["output_tokens_1"] as number | undefined) ?? 0,
    };
    const metrics2 = {
      model: "_default",
      inputTokens: (action["input_tokens_2"] as number | undefined) ?? 0,
      outputTokens: (action["output_tokens_2"] as number | undefined) ?? 0,
    };
    try {
      await manager.deductTeam(action["team_id"] as string, action["user_id"] as string, metrics1);
    } catch {
      // ignore first deduct failure
    }
    try {
      await manager.deductTeam(action["team_id"] as string, action["user_id"] as string, metrics2);
      return { second_deduct_error: null };
    } catch (e) {
      return { second_deduct_error: errorCode(e) };
    }
  }

  if (op === "deduct_then_refund") {
    const metrics = {
      model: "_default",
      inputTokens: (action["input_tokens"] as number | undefined) ?? 0,
      outputTokens: (action["output_tokens"] as number | undefined) ?? 0,
    };
    try {
      const ded = await manager.deduct(action["user_id"] as string, metrics);
      const ref = await manager.refundCredits(ded.transactionId);
      return {
        balance_after_refund: q(ref.newBalance),
        error: null,
      };
    } catch (e) {
      return { error: errorCode(e) };
    }
  }

  if (op === "evaluate_expr") {
    const vars = action["vars"] as Record<string, number>;
    const decVars: Record<string, Decimal> = {};
    for (const [k, v] of Object.entries(vars)) {
      decVars[k] = new Decimal(v);
    }
    try {
      const result = evaluateExpression(action["expr"] as string, decVars);
      return { result: q(result), error: null };
    } catch (e) {
      return { error: errorCode(e) };
    }
  }

  throw new Error(`Unknown action op: ${op}`);
}

// ---------------------------------------------------------------------------
// Vitest test suite
// ---------------------------------------------------------------------------

describe("Cross-SDK billing parity (JS side)", () => {
  for (const scenario of SCENARIOS) {
    const name = scenario["name"] as string;
    it(name, async () => {
      const result = await runAction(scenario, PRICING);
      const expected = scenario["assert"] as Record<string, unknown>;
      assertFields(name, expected, result);
    });
  }
});

File: tests/parity/run_parity.py
"""Cross-SDK billing parity runner — Python side.

Loads billing_scenarios.json and exercises each scenario against the Python
MemoryStore+CreditManager. Results are saved to parity_results_py.json so the
JS runner can diff them, and the file is also used as a standalone pytest suite.

Run directly:  python tests/parity/run_parity.py
Run via pytest: pytest tests/parity/run_parity.py -v
"""

from __future__ import annotations

import json
import sys
from decimal import Decimal
from pathlib import Path

import pytest

PARITY_DIR = Path(__file__).parent
SCENARIOS_FILE = PARITY_DIR / "billing_scenarios.json"

# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _load_scenarios() -> dict:
    with open(SCENARIOS_FILE) as f:
        return json.load(f)


def _make_manager(pricing_cfg: dict, billing_mode: str = "strict", overdraft_floor: Decimal | None = None, min_balance: Decimal | None = None):
    """Build a CreditManager with MemoryStore pre-loaded with pricing."""
    from ducto import CreditManager
    from ducto.interface.memory import MemoryStore

    store = MemoryStore()
    kwargs: dict = {"store": store}
    if billing_mode == "overdraft":
        kwargs["policy"] = "overdraft"
        if overdraft_floor is not None:
            kwargs["overdraft_floor"] = overdraft_floor
    else:
        kwargs["policy"] = "strict_prepaid"

    m = CreditManager(**kwargs)
    m.publish_pricing_from_dict(pricing_cfg)
    return m, store


def _run_setup(store, manager, steps: list[dict]) -> None:
    from ducto.interface.models import CreditMetadata

    for step in steps:
        op = step["op"]
        if op == "add_credits":
            store.add_credits(step["user_id"], Decimal(step["amount"]))
        elif op == "set_user_plan":
            store.set_user_plan(step["user_id"], step["plan_key"])
        elif op == "create_team":
            # MemoryStore generates a UUID for team_id; patch both _teams and
            # _team_members to use the fixture's stable id for predictable tests.
            result = store.create_team(step["team_id"], Decimal(step["initial_balance"]))
            actual_id = result.team_id
            if actual_id != step["team_id"]:
                team_rec = store._teams.pop(actual_id)
                team_rec.id = step["team_id"]
                store._teams[step["team_id"]] = team_rec
                store._team_members[step["team_id"]] = store._team_members.pop(actual_id)
        elif op == "add_team_member":
            cap = Decimal(step["spend_cap"]) if step.get("spend_cap") else None
            store.add_team_member(step["team_id"], step["user_id"], spend_cap=cap)
        else:
            raise ValueError(f"Unknown setup op: {op}")


def _run_action(scenario: dict, pricing_cfg: dict) -> dict:
    """Execute the scenario action and return a normalised result dict."""
    from ducto import CreditManager, UsageMetrics
    from ducto.expr import evaluate_expression
    from ducto.interface.memory import MemoryStore

    action = scenario["action"]
    op = action["op"]

    # Determine billing mode from action or default
    billing_mode = action.get("billing_mode", "strict")
    overdraft_floor_raw = action.get("overdraft_floor")
    overdraft_floor = Decimal(overdraft_floor_raw) if overdraft_floor_raw is not None else None
    min_balance_raw = action.get("min_balance")
    min_balance_override = Decimal(min_balance_raw) if min_balance_raw is not None else None

    manager, store = _make_manager(
        pricing_cfg,
        billing_mode=billing_mode,
        overdraft_floor=overdraft_floor,
    )

    _run_setup(store, manager, scenario.get("setup", []))

    if op == "deduct":
        metrics = UsageMetrics(
            model=action.get("model", "_default"),
            input_tokens=action.get("input_tokens", 0),
            output_tokens=action.get("output_tokens", 0),
        )
        try:
            res = manager.deduct(action["user_id"], metrics)
            return {
                "balance_after": str(res.balance_after),
                "allowance_consumed": str(res.allowance_consumed),
                "plan_covered": getattr(res, "plan_covered", None),
                "error": res.error,
            }
        except Exception as e:
            return {"error": _error_code(e)}

    elif op == "deduct_idempotent":
        metrics = UsageMetrics(
            model=action.get("model", "_default"),
            input_tokens=action.get("input_tokens", 0),
            output_tokens=action.get("output_tokens", 0),
        )
        try:
            res1 = manager.deduct(action["user_id"], metrics, idempotency_key=action["idempotency_key"])
            res2 = manager.deduct(action["user_id"], metrics, idempotency_key=action["idempotency_key"])
            return {
                "balance_after": str(res1.balance_after),
                "idempotent_on_replay": res2.idempotent,
                "error": None,
            }
        except Exception as e:
            return {"error": _error_code(e)}

    elif op == "get_user_plan":
        res = store.get_user_plan(action["user_id"])
        return {
            "plan_id": res.plan_id,
            "billing_mode": res.default_billing_mode,
        }

    elif op == "lease_reserve_settle":
        user_id = action["user_id"]
        reserve_amt = Decimal(action["reserve_amount"])
        settle_amt = Decimal(action["settle_amount"])
        try:
            lease = manager.reserve(user_id, reserve_amt)
            min_bal = min_balance_override or (manager._engine.min_balance if manager._engine else Decimal(0))
            res = manager._store.settle_lease(
                user_id, lease.lease_id, settle_amt,
                min_balance=min_bal,
            )
            return {"balance_after": str(res.balance_after), "error": res.error}
        except Exception as e:
            return {"error": _error_code(e)}

    elif op == "lease_reserve_release":
        user_id = action["user_id"]
        reserve_amt = Decimal(action["reserve_amount"])
        try:
            lease = manager.reserve(user_id, reserve_amt)
            manager.release(user_id, lease.lease_id)
            available = manager.get_available(user_id)
            return {"available_after_release": str(available.available), "error": None}
        except Exception as e:
            return {"error": _error_code(e)}

    elif op == "team_deduct":
        metrics = UsageMetrics(
            model=action.get("model", "_default"),
            input_tokens=action.get("input_tokens", 0),
            output_tokens=action.get("output_tokens", 0),
        )
        try:
            res = manager.deduct_team(action["team_id"], action["user_id"], metrics)
            return {"team_balance_after": str(res.team_balance_after), "error": res.error}
        except Exception as e:
            return {"error": _error_code(e)}

    elif op == "team_deduct_twice":
        metrics1 = UsageMetrics(
            model="_default",
            input_tokens=action.get("input_tokens_1", 0),
            output_tokens=action.get("output_tokens_1", 0),
        )
        metrics2 = UsageMetrics(
            model="_default",
            input_tokens=action.get("input_tokens_2", 0),
            output_tokens=action.get("output_tokens_2", 0),
        )
        try:
            manager.deduct_team(action["team_id"], action["user_id"], metrics1)
        except Exception:
            pass
        try:
            manager.deduct_team(action["team_id"], action["user_id"], metrics2)
            return {"second_deduct_error": None}
        except Exception as e:
            return {"second_deduct_error": _error_code(e)}

    elif op == "deduct_then_refund":
        metrics = UsageMetrics(
            model="_default",
            input_tokens=action.get("input_tokens", 0),
            output_tokens=action.get("output_tokens", 0),
        )
        try:
            ded = manager.deduct(action["user_id"], metrics)
            ref = manager.refund_credits(ded.transaction_id)
            return {"balance_after_refund": str(ref.new_balance), "error": None}
        except Exception as e:
            return {"error": _error_code(e)}

    elif op == "evaluate_expr":
        from ducto.expr import evaluate_expression
        from ducto.engine import _q
        try:
            result = evaluate_expression(action["expr"], action["vars"])
            return {"result": str(_q(result)), "error": None}
        except Exception as e:
            return {"error": _error_code(e)}

    else:
        raise ValueError(f"Unknown action op: {op}")


def _error_code(exc: Exception) -> str:
    """Convert an exception to a short code matching the scenario's expected error."""
    name = type(exc).__name__
    mapping = {
        "InsufficientCreditsError": "insufficient_credits",
        "CapReachedError": "cap_reached",
        "InsufficientTeamBalanceError": "insufficient_team_balance",
        "SpendCapExceededError": "spend_cap_exceeded",
        "LeaseExpiredError": "lease_expired",
        "LeaseNotFoundError": "lease_not_found",
    }
    # Prefer explicit error code attribute.
    code = getattr(exc, "code", None) or getattr(exc, "error", None)
    if code:
        return str(code)
    # Some exceptions embed the specific code in the message (e.g. InsufficientCreditsError
    # raised for team deducts includes "spend_cap_exceeded" in the message text).
    msg = str(exc)
    for known_code in ("spend_cap_exceeded", "cap_reached", "insufficient_team_balance",
                       "lease_expired", "lease_not_found", "invalid_amount"):
        if known_code in msg:
            return known_code
    return mapping.get(name, name)


def _assert_scenario(scenario: dict, result: dict) -> list[str]:
    """Return a list of failure messages (empty = pass)."""
    failures = []
    expected = scenario.get("assert", {})
    for key, exp_val in expected.items():
        got = result.get(key)
        # Normalise Decimal strings for comparison
        if isinstance(exp_val, str) and "." in exp_val:
            try:
                exp_d = Decimal(exp_val)
                got_d = Decimal(str(got)) if got is not None else None
                if got_d is None or exp_d != got_d:
                    failures.append(f"  {key}: expected {exp_val!r}, got {got!r}")
                continue
            except Exception:
                pass
        if exp_val != got:
            failures.append(f"  {key}: expected {exp_val!r}, got {got!r}")
    return failures


# ---------------------------------------------------------------------------
# pytest parametrize
# ---------------------------------------------------------------------------

_DATA = _load_scenarios()
_PRICING = _DATA["pricing_config"]
_SCENARIOS = _DATA["scenarios"]


@pytest.mark.parametrize("scenario", _SCENARIOS, ids=[s["name"] for s in _SCENARIOS])
def test_billing_scenario(scenario: dict) -> None:
    result = _run_action(scenario, _PRICING)
    failures = _assert_scenario(scenario, result)
    if failures:
        pytest.fail(f"Scenario '{scenario['name']}' failed:\n" + "\n".join(failures) + f"\n  result: {result}")


# ---------------------------------------------------------------------------
# CLI entry point — writes parity_results_py.json for cross-runner diffing
# ---------------------------------------------------------------------------

if __name__ == "__main__":
    data = _load_scenarios()
    pricing = data["pricing_config"]
    scenarios = data["scenarios"]

    results = {}
    errors = []
    for sc in scenarios:
        result = _run_action(sc, pricing)
        failures = _assert_scenario(sc, result)
        results[sc["name"]] = result
        if failures:
            errors.append(f"FAIL {sc['name']}:\n" + "\n".join(failures))
        else:
            print(f"  PASS {sc['name']}")

    out_file = PARITY_DIR / "parity_results_py.json"
    with open(out_file, "w") as f:
        json.dump(results, f, indent=2, default=str)
    print(f"\nResults written to {out_file}")

    if errors:
        print("\nFAILURES:")
        for e in errors:
            print(e)
        sys.exit(1)
    else:
        print(f"\nAll {len(scenarios)} scenarios passed.")
